File: lab1/expr.pl
#!/usr/bin/perl -w
%ops = qw#+ Add * Mult / Div - Sub#; #robimy sobie hash do latwego wypisywania
sub Add ($$) { return [shift,"+",shift]}; #tworzenie listy zawierajacej referencje do lewego syna (lub wartosc, jezeli to lisc)
sub Mult ($$) { return [shift,"*",shift]}; #latwo w razie czego opakowac obiektowo
sub Div ($$) { return [shift,"/",shift]};
sub Sub ($$) { return [shift,"-",shift]};
sub Minus {["-",shift]}; #dla minusa specjalny rodzaj listy
sub myeval { #nazwa eval juz zajeta, ponadto znacznie ulatwia parsing do drzewa
	my $_=pop; #argument funkcji
	if(ref() && @$_[0] eq "-"){ #jezeli referencja i lista Minusa
		return eval '- myeval(@$_[1])' #ewaluujemy przy uzyciu zwyklej arytmetyki perla
	} elsif (ref()) { #jezeli referencja i normalny przypadek
		return eval'myeval(@$_[0])'.@$_[1].'myeval(@$_[2])' #znowu arytmetyka perla
	} else { 
		return $_
	}
}	
sub simplify_expr {
	my $_=pop;
	if(ref() && @$_[0] eq "-"){ #referencja Minusowa
		@$_[1]=simplify_expr(@$_[1]); #upraszczamy podwyrazenie
		return $_;
	} elsif(ref()) { #referencja
		my ($a,$b)=(simplify_expr(@$_[0]),simplify_expr(@$_[2])); #upraszczamy podwyrazenia
		if(@$_[1] eq '+' && $a==0){# 0+e
			return $b; # e
		} elsif(@$_[1] eq '*' && $a==1){# 1*e
			return $b; # e
		} elsif (@$_[1] eq '+' && $b==0){# e+0
			return $a; # e
		} elsif (@$_[1] eq '*' && $b==1){# e*1
			return $a; # e
		} else { #normalny przypadek
			return [$a,@$_[1],$b] #zwracamy poddrzewo
		} 
	} else { #wartosc
		return $_
	} 
} 
sub string { 
	my $_=pop;
	if(ref() && @$_[0] eq "-") {
		return "Minus(".string(@$_[1]).")"
	} elsif (ref()){
		return $ops{@$_[1]}."(".string(@$_[0]).", ".string(@$_[2]).")"
	} else {
		return $_
	}
}
print "Test dla Minus(Add(Div(Mult(2,Sub(10,Add(0,Mult(2,3)))),3),1))...\n";
$ref = eval "Minus(Add(Div(Mult(2,Sub(10,Add(0,Mult(2,3)))),3),1))";
print myeval($ref)."\n";
print string(simplify_expr($ref))."\n";

print "Podaj wyrazenie...\n>";
while($line = <>){
	last if !$line;
	$ref=eval $line;
	if($@){
		warn "error parsing input!"
	} else {
		print "Wynik: ".myeval($ref)."\n";
		print "Uproszczony: ".string(simplify_expr($ref))."\n>";
	}
}

File: lab1/expr.ml
type expr = 
        | Int of int
        | Float of float
        | Add of expr * expr
        | Sub of expr * expr
        | Mult of expr * expr
        | Div of expr * expr
        | Minus of expr

let rec eval expr = match expr with
        | Int a -> float a
        | Float a -> a
        | Add (a,b) -> (eval a)+.(eval b)
        | Sub (a,b) -> (eval a)-.(eval b)
        | Mult (a,b) -> (eval a)*.(eval b)
        | Div (a,b) -> (eval a)/.(eval b)
        | Minus a -> -.(eval a)
let rec simplify expr = match expr with
        | Int a -> Float (float a)
        | Float a -> Float a
        | Minus a -> if (simplify a)=Float 0. then
                Float 0. else Minus (simplify a)
        | Div (a,b) -> if(simplify b)=Float 1. then
                simplify a else Div (simplify a, simplify b)
        | Add (a,b) -> if (simplify a)=Float 0. then 
                simplify b else if (simplify b)=Float 0. then
                simplify a else Add (simplify a,simplify b)
        | Sub (a,b) -> if (simplify a)=Float 0. then 
                Minus (simplify b) else if (simplify b)=Float 0. then
                simplify a else Sub (simplify a,simplify b)
        | Mult (a,b) -> if(simplify a)=Float 1. then
                simplify b else if (simplify b)=Float 1. then
                simplify a else Mult (simplify a, simplify b)

let rec simplify_expr expr= 
        let expr2 = simplify expr in if expr2 <> expr then simplify_expr expr2 else expr2

let print_expr expr =
        let openpar prec op_prec =
                if prec > op_prec then "(" else "" in
        let closepar prec op_prec =
                if prec > op_prec then ")" else "" in
        let rec print prec expr = 
                match expr with
                Float f -> string_of_float f
                | Int i -> string_of_int i
                | Add(f, g) ->
                        openpar prec 1 ^ print 1 f ^ "+" ^ print 1 g ^ closepar prec 1
                | Sub(f, g) ->
                        openpar prec 1 ^ print 1 f ^ "-" ^ print 2 g ^ closepar prec 1
                | Mult(f, g) ->
                        openpar prec 3 ^ print 3 f ^ "*" ^ print 3 g ^ closepar prec 3
                | Div(f, g) ->
                        openpar prec 3 ^ print 3 f ^ "/" ^ print 4 g ^ closepar prec 3
                | Minus(a) ->
                        match a with
                                Float f -> openpar prec 0 ^ "-" ^ print 0 a ^ closepar prec 0
                                | Int i -> openpar prec 0 ^ "-" ^ print 0 a ^ closepar prec 0
                                | _ -> openpar prec 0 ^ "-" ^ openpar prec 0 ^ print 0 a ^ closepar prec 0 ^ closepar prec 0
        in print 0 expr
